src/gpuarray_blas_cuda_cublas.c

```diff
@@ -295,44 +287,80 @@ static int sgemmBatch(cb_order order, cb_transpose transA, cb_transpose transB,
     offB = lt;
   }
 
-  cuda_enter(ctx);
-
-  for (i = 0; i < batchCount; i++) {
-    ASSERT_BUF(A[i]);
-    ASSERT_BUF(B[i]);
-    ASSERT_BUF(C[i]);
-    cuda_wait(A[i], CUDA_WAIT_READ);
-    cuda_wait(B[i], CUDA_WAIT_READ);
-    cuda_wait(C[i], CUDA_WAIT_READ|CUDA_WAIT_WRITE);
-    A_l[i] = ((float *)A[i]->ptr) + offA[i];
-    B_l[i] = ((float *)B[i]->ptr) + offB[i];
-    C_l[i] = ((float *)C[i]->ptr) + offC[i];
-  }
-
-  cuMemAlloc(&Ta, sizeof(float *) * batchCount * 3);
-  Aa = Ta;
-  Ba = Ta + (batchCount * sizeof(float *));
-  Ca = Ta + (batchCount * sizeof(float *) * 2);
-
-  cuMemcpyHtoD(Ta, T_l, sizeof(float *) * batchCount * 3);
-
-  err = cublasSgemmBatched(ctx->blas_handle, convT(transA), convT(transB),
-                           M, N, K, &alpha, (const float **)Aa, lda,
-                           (const float **)Ba, ldb, &beta,
-                           (float **)Ca, ldc, batchCount);
-  cuMemFree(Ta);
-  if (err != CUBLAS_STATUS_SUCCESS) {
-    cuda_exit(ctx);
-    if (err == CUBLAS_STATUS_ARCH_MISMATCH)
-      return GA_DEVSUP_ERROR;
-    return GA_BLAS_ERROR;
-  }
+  // use parallel cublasSgemm calls rather than cublasSgemmBatched for large products
+  // (compute products in double because they can be large and we don't need to be exact)
+  const double threshold = 650;
```

Comment:
64-bit integer (or size_t) will be big enough.  Otherwise the array wouldn't fit on the GPU.